webglframework.js

```diff
@@ -1,523 +1,480 @@
-      gl.bufferData(gl.ARRAY_BUFFER, new Float32Array([  // Vertex coordinates
-         1.0, 1.0, 1.0,  -1.0, 1.0, 1.0,  -1.0,-1.0, 1.0,   1.0,-1.0, 1.0, // front
-         1.0, 1.0, 1.0,   1.0,-1.0, 1.0,   1.0,-1.0,-1.0,   1.0, 1.0,-1.0, // right
-         1.0, 1.0, 1.0,   1.0, 1.0,-1.0,  -1.0, 1.0,-1.0,  -1.0, 1.0, 1.0, // up
-        -1.0, 1.0, 1.0,  -1.0, 1.0,-1.0,  -1.0,-1.0,-1.0,  -1.0,-1.0, 1.0, // left
-        -1.0,-1.0,-1.0,   1.0,-1.0,-1.0,   1.0,-1.0, 1.0,  -1.0,-1.0, 1.0, // down
-         1.0,-1.0,-1.0,  -1.0,-1.0,-1.0,  -1.0, 1.0,-1.0,   1.0, 1.0,-1.0  // back
-      ]), gl.STATIC_DRAW);
-      var a = gl.getAttribLocation(W.P, "position");
-      gl.vertexAttribPointer(a, 3, gl.FLOAT, false, 0, 0);
-      gl.enableVertexAttribArray(a);
-
-      // Set indices
-      gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, gl.createBuffer());
-      gl.bufferData(gl.ELEMENT_ARRAY_BUFFER,  new Uint16Array([
-        //0, 1, 2,   0, 2, 3,  // front
-        4, 5, 6,   4, 6, 7,  // right
-        8, 9, 10,  8, 10,11, // up
-        //12,13,14,  12,14,15, // left
-        //16,17,18,  16,18,19, // down
-        //20,21,22,  20,22,23  // back
-      ]), gl.STATIC_DRAW);
-      var n = 12;
-
-      // render
-      //gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
-      gl.drawElements(gl.TRIANGLES, n, gl.UNSIGNED_SHORT, 0);
-    */
-
-    
-    
-    
-    
-    
-    
-      // Draw
-      // gl.drawArrays(gl.TRIANGLES, 0, gl.getBufferParameter(gl.ARRAY_BUFFER, gl.BUFFER_SIZE)/8);
-      gl.drawArrays(4, 0, gl.getBufferParameter(34962, 34660)/8);
-      
+
+      // Use a renderer (triangles by default)
+      W.renderers[object.r || 'triangles'](object);
     }
-    
   },
   
+  // Helpers
+  // -------
+  
+  // Interpolate a property between two values
+  lerp: (item, property) => 
+    W.next[item]?.t
+    ? W.current[item][property] + (W.next[item][property] -  W.current[item][property]) * (W.next[item].f / W.next[item].t)
+    : W.next[item][property],
+  
+  // Transition an item
+  transition: (item, m = new DOMMatrix) =>
+    W.next[item]
+    ? m
+      .translateSelf(W.lerp(item, 'x'), W.lerp(item, 'y'), W.lerp(item, 'z'))
+      .rotateSelf(W.lerp(item, 'rx'),W.lerp(item, 'ry'),W.lerp(item, 'rz'))
+      .scaleSelf(W.lerp(item, 'w'),W.lerp(item, 'h'),W.lerp(item, 'd'))
+    : m,
+    
   // Compute the distance squared between two objects (useful for sorting transparent items)
-  dist: (a, b) => (b.m41 - a.m41)**2 + (b.m42 - a.m42)**2 + (b.m43 - a.m43)**2
-}
-
-// When everything is loaded: setup WebGL program, light, camera, action
-W.reset();
-W.s();
-W.light({z:1});
-W.camera({});
-W.d();
+  dist: (a, b = W.next.camera) => a && b ? (b.m.m41 - a.m.m41)**2 + (b.m.m42 - a.m.m42)**2 + (b.m.m43 - a.m.m43)**2 : 0,
+  
+  // Built-in objects
+  // ----------------
+  
+  group: t => W.setState(t, 'group'),
+  
+  move: (t, delay) => setTimeout(()=>{ W.setState(t) }, delay || 1),
```

Comment:
Maybe use `requestAnimationFrame` here?

Reply:
er, I'd like to but is it possible to set up a delay (ex: 3000ms) using requestAnimationFrame  ? :)

Reply:
true, we already count frames for the transitions, so we can probavly implement the delays with the same system (I'm still thinking how to achieve it)

And I didn't forget the transition time as second parameter, not sure it'll save a lot but I'll try it !